gapis/api/vulkan/looping_vulkan_control_flow_generator.go

```diff
@@ -338,209 +352,260 @@ func newFrameLoop(ctx context.Context, graphicsCapture *capture.GraphicsCapture,
+						{
+							// Write out some custom bytecode for the loop.
+							stateBuilder.write(stateBuilder.cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
+								f.loopCountPtr = b.AllocateMemory(4)
+								b.Push(value.S32(f.loopCount - 1))
+								b.Store(f.loopCountPtr)
+								b.JumpLabel(uint32(0x1))
+								return nil
+							}))
+						}
+
+						// Do first N-1 iterations of mid-loop stuff.
+						if err := f.writeLoopContents(ctx, f.out); err != nil {
+							return err
+						}
+
+						// Do state rewind stuff.
+						{
+							// Now we need to emit the instructions to reset the state, before the conditional branch back to the start of the loop.
+							if err := f.resetResources(ctx, stateBuilder); err != nil {
+								return fmt.Errorf("FrameLoop: Failed to reset changed resources %v.", err)
+							}
+						}
+
+						// Write out the conditional jump to the start of the state rewind code to provide the actual looping behaviour
+						{
+							stateBuilder.write(stateBuilder.cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
+								b.Load(protocol.Type_Int32, f.loopCountPtr)
+								b.Sub(1)
+								b.Clone(0)
+								b.Store(f.loopCountPtr)
+								b.JumpNZ(uint32(0x1))
+								return nil
+							}))
+						}
+					}
 
-			// Finally, we've done all the processing for a loop. Nothing left to do.
-			return nil
+					// Start the perfetto
+					{
+						cmds := make([]api.Cmd, 0)
+						cmds = append(cmds, f.createVkDeviceWaitIdleCommandsForDevices(ctx, f.out.State())...)
+						waitForFenceCmd := f.createWaitForFence(ctx, uint32(id.GetID()), func(ctx context.Context, request *gapir.FenceReadyRequest) {
+							f.loopCallbacks.preLoop(ctx, request)
+						})
+						cmds = append(cmds, waitForFenceCmd)
+
+						if err := f.buildCommands(ctx, cmdId, cmds, f.out); err != nil {
+							return err
+						}
+					}
 
-		} else { // We're currently inside the loop.
+					// Do the final iteration of mid-loop stuff.
+					if err := f.writeLoopContents(ctx, f.out); err != nil {
+						return err
+					}
 
-			// Lets just remember the command we've seen so we can do all the work we need at the end of the loop.
-			// This is done because the information we need to transform the loop is only available at that time;
-			// due to the possibility of preceeding transforms modifing the loop contents in-flight.
-			f.capturedLoopCmds = append(f.capturedLoopCmds, cmd)
-			f.capturedLoopCmdIds = append(f.capturedLoopCmdIds, cmdId)
+					// End the perfetto
+					{
+						cmds := make([]api.Cmd, 0)
+						cmds = append(cmds, f.createVkDeviceWaitIdleCommandsForDevices(ctx, f.out.State())...)
 
-			log.D(ctx, "FrameLoop: inside loop at frame %v, cmdId %v, cmd %v.", f.frameNum, cmdId, cmd)
+						fenceID := uint32(0x3ffffff)
+						waitForFenceCmd := f.createWaitForFence(ctx, fenceID, func(ctx context.Context, request *gapir.FenceReadyRequest) {
```

Comment:
This should definitely be in a different place possibly wait_for_perfetto. Framelooping and connecting to perfetto are entirely two different concept and they should be separated. In this way they are completely entangled together and there is no way to separate them.